das/state.go

```diff
@@ -93,67 +107,76 @@ func (s *coordinatorState) updateHead(newHead int64) {
 	s.checkDone()
 }
 
-func (s *coordinatorState) newRecentJob(header *header.ExtendedHeader) job {
+// recentJob creates a job to process recently produced header
+func (s *coordinatorState) recentJob(header *header.ExtendedHeader) job {
 	height := uint64(header.Height())
 	// move next, to prevent catchup job from processing same height
 	if s.next == height {
 		s.next++
 	}
 	s.nextJobID++
 	return job{
-		id:             s.nextJobID,
-		isRecentHeader: true,
-		header:         header,
-		From:           height,
-		To:             height,
+		id:     s.nextJobID,
+		Type:   recentJob,
+		header: header,
+		From:   height,
+		To:     height,
 	}
 }
 
-// nextJob will return header height to be processed and done flag if there is none
+// nextJob will return next job according to priority (catchup -> retry)
 func (s *coordinatorState) nextJob() (next job, found bool) {
-	// all headers were sent to workers.
-	if s.next > s.networkHead {
-		return job{}, false
-	}
-
-	// try to take from retry first
-	if next, found := s.nextFromRetry(); found {
-		return next, found
+	// check for catchup job
+	if job, found := s.catchupJob(); found {
+		return job, found
 	}
 
-	j := s.newJob(s.next, s.networkHead)
+	// if all headers were tried already, make a retry job
+	return s.retryJob()
+}
 
-	s.next += s.samplingRange
+// catchupJob creates a catchup job if catchup is not finished
+func (s *coordinatorState) catchupJob() (next job, found bool) {
 	if s.next > s.networkHead {
-		s.next = s.networkHead + 1
+		return job{}, false
 	}
 
+	to := s.next + s.samplingRange - 1
+	if to > s.networkHead {
+		to = s.networkHead
+	}
+	j := s.newJob(catchupJob, s.next, to)
+	s.next = to + 1
 	return j, true
 }
 
-func (s *coordinatorState) nextFromRetry() (job, bool) {
-	if len(s.retry) == 0 {
-		return job{}, false
-	}
+// retryJob creates a job to retry previously failed header
+func (s *coordinatorState) retryJob() (next job, found bool) {
+	for h, count := range s.failed {
+		// TODO(@walldiss): limit max amount of retries until retry backoff is implemented
```

Comment:
pls make issue for this if you haven't already. I think that we should handle backoff a bit differently bc if you retry the height 3+ times and you still can't sample it, that means something's pretty wrong.

Reply:
There is an issue for that https://github.com/celestiaorg/celestia-node/issues/2027